examples/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,447 @@
+"""
+Title: Classification with Neural Decision Forests
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/15
+Last modified: 2021/01/15
+Description: How to train differentiable decision trees for end-to-end learning in deep neural networks.
+"""
+
+"""
+## Introduction
+
+This example provides an implementation of the
+[Deep Neural Decision Forest](https://ieeexplore.ieee.org/document/7410529)
+model, introduced by P. Kontschieder et al. for structured data classification.
+This model introduces a stochastic and differentiable decision tree model,
+trained end-to-end, unifying decision trees with deep representation learning.
+
+## The dataset
+
+This example uses the
+[United States Census Income Dataset](https://archive.ics.uci.edu/ml/datasets/census+income)
+provided by the
+[UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php).
+The task is binary classification
+to predict whether a person is likely to be making over USD 50k a year.
+
+The dataset includes 48,842 instances with 14 input features: 5 numerical features
+and 9 categorical features.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+from tensorflow import keras
+from tensorflow.keras import layers
+import math
+
+"""
+## Prepare the data
+"""
+
+CSV_HEADER = [
+    "age",
+    "workclass",
+    "fnlwgt",
+    "education",
+    "education_num",
+    "marital_status",
+    "occupation",
+    "relationship",
+    "race",
+    "gender",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "native_country",
+    "income_bracket",
+]
+
+train_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+)
+train_data = pd.read_csv(train_data_url, header=None, names=CSV_HEADER)
+
+test_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test"
+)
+test_data = pd.read_csv(test_data_url, header=None, names=CSV_HEADER)
+
+print(f"Train dataset shape: {train_data.shape}")
+print(f"Test dataset shape: {test_data.shape}")
+
+"""
+We remove the first record as it is not a valid data example, and we remove a trailing
+'dot' in the class labels.
```

Comment:
Suggestion: rephrase a little to avoid repetition ("remove...remove")

```suggestion
Remove the first record (because it is not a valid data example) and a trailing
'dot' in the class labels.
```